### PR TITLE
Paid Blocks Nudge: Fix Border by Injecting has-warning Style

### LINKED ISSUE #13007: Premium block flows: border/spacing adjustments
A list of small style fixes we should do — follow up from https://github.com/Automattic/jetpack/pull/12963:

- [ ] Border/margin/padding funkiness on hover:
    Screenshot 2019-07-05 at 15 56 01

- [ ] And the banner seems to "jump" visually ~1-2px when selecting the block so it doesn't feel as smooth as it does for "missing block" and similar core alerts. Basically, you see some movement in the block when selecting it, while you don't see this movement with other blocks.

### LINKED ISSUE #13008: Premium block nudge: turn banner blue when multiple blocks are selected
Via https://github.com/Automattic/jetpack/pull/12963#issuecomment-508735110

> >   I'd expect upgrade banner to get a blue hue when multiple blocks are selected:
> >     image
> >   — might be just a matter of correct z-index or `background:transparent`?
> 
> Ideally, yeah. However, I've seen a bunch of code and/or styling in core (e.g. in the `Warning` component that this PR uses) that says something like 'make sure it works with multi-block selection' (or something like that), so it might be a tad more involved :grimacing: 
> 
> Edit: E.g. https://github.com/WordPress/gutenberg/blob/21e2dce2b287347e21c3e0618bc8fa8c0c1ef5ca/packages/block-editor/src/components/warning/style.scss#L11-L14 Though that's really just `background:transparent` :joy:

### PR DESCRIPTION
**Pull request by ockham**
Fixes #13007.
Fixes #13008, too :sunglasses: 

#### Changes proposed in this Pull Request:

Paid Blocks Nudge: Fix Border by Injecting has-warning Style

Prior to this PR, we used custom styling for the `jetpack-paid-block__wrapper` class, using padding and negative margins. This lead to visual glitches such as #13007:

Screenshot 2019-07-05 at 15 56 01

As [documented there](https://github.com/Automattic/jetpack/issues/13007#issuecomment-510009903), this is because that kind of styling doesn't play well with Gutenberg's own, which is using _absolute_ positioning on `:before` selectors such as

-  `.block-editor-block-list__layout .block-editor-block-list__block .block-editor-block-list__block-edit:before` 
- `.block-editor-block-list__layout .block-editor-block-list__block.is-selected>.block-editor-block-list__block-edit:before`
- `.block-editor-block-list__layout .block-editor-block-list__block.has-warning .block-editor-block-list__block-edit:before`

So ideally, we want to piggy-back on those selectors -- the `has-warning` ones in particular, since they do precisely what we want (also on a conceptual level). However, that's not easy from within the block, since those selectors apply to the block _wrapper_. Class names for that wrapper are pretty much [hard-wired](https://github.com/WordPress/gutenberg/blob/eba2fc8ba1ffbd5997cf47b3aa7002f8a5597e9f/packages/block-editor/src/components/block-list/block.js#L392-L408), and there aren't really any props to that element that we could (or should) modify to enforce the styling we want.

But fortunately, there's a [filter](https://developer.wordpress.org/block-editor/developers/filters/block-filters/#editor-blocklistblock) to modify those wrappers :tada: 

This PR thus uses that filter to inject the `has-warning` class, fixing #13007, and allowing us to drop a bunch of custom styling.

However, the `has-warning` selector also adds styling to prevent user interactivity with the given block (see https://github.com/Automattic/jetpack/pull/13019#issuecomment-510129941 for details), so we need to override that behavior, since we want to ensure the user can continue to interact with the block. To that end, we add another class name, `is-interactive` for increase specificity, and override the styling that would otherwise prevent user interactions. 

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
Modifies the existing paid plan nudge.

#### Testing instructions:

- Switch to this branch locally, build Jetpack (`yarn build`), and start Docker (`yarn docker:up`)
- Now add the following line to any file in `docker/mu-plugins`, e.g. a newly created `0-blocks-upgrade.php` (needs to start with `<?php`): `define ( 'JETPACK_SHOW_BLOCK_UPGRADE_NUDGE', true );`
- Make sure you're connected to WP.com, and on a free plan
- Start a new post, and insert the 'Simple Payments' block
- Select the block, and enter some information into its text fields.
- Unselect the block again
- Verify that the visual glitch seen in the above screenshot is fixed.
- Verify that you can still select and edit the block.
- Finally, verify that #13008 is also fixed by this PR :tada: 

#### Proposed changelog entry for your changes:

(Covered by whatever we add for #12823)

### CODE REVIEW

**Review by simison — DISMISSED** [review]
I'm now getting this class for every block:
image

Furthermore, I cannot edit fields in simple payments block which defeats the purpose of us letting users to try it out.

**Comment by ockham** [comment]
Ugh my bad. I'll try to find a fix...

File: extensions/shared/register-jetpack-block.js [hunk 1]
```diff
@@ -19,6 +21,13 @@ function requiresPlan( unavailableReason, details ) {
 	return false;
 }
 
+// Injecting the `has-warning` class into the block wrapper component gives us
+// the right kind of borders around the block, both visually and conceptually.
+const withHasWarningClassName = createHigherOrderComponent(
+	BlockListBlock => props => <BlockListBlock { ...props } className="has-warning" />,
```
Comment (simison) [inline extensions/shared/register-jetpack-block.js]: This will need block-name check before applying the class.

**Comment by simison** [comment]
Might be easiest to keep this implementation and override the background color + layer that stops pointer interactions?

**Comment by ockham** [comment]
> Might be easiest to keep this implementation and override the background color + layer that stops pointer interactions?

Exactly my plan :+1:

File: extensions/shared/register-jetpack-block.js (hunk 1 above)
Comment (ockham) [reply extensions/shared/register-jetpack-block.js]: :+1: dc630d9

**Comment by ockham** [comment]
More fun:

The relevant styling isn't only https://github.com/WordPress/gutenberg/blob/c5e8784799f4e0f4d0891f52fae95375f43b76f7/packages/block-editor/src/components/block-list/style.scss#L219-L224

but also https://github.com/WordPress/gutenberg/blob/c5e8784799f4e0f4d0891f52fae95375f43b76f7/packages/block-editor/src/components/block-list/style.scss#L252-L261 which generates a transparent pseudo-element that eats up cursor focus, and tints its what's behind it grey.

File: extensions/shared/with-has-warning-is-interactive-class-names/style.scss [hunk 2]
```diff
@@ -0,0 +1,12 @@
+.block-editor-block-list__layout
```
Comment (simison) [inline extensions/shared/with-has-warning-is-interactive-class-names/style.scss]: Might be worth noting with a comment why these overrides exist. Not a blocker for merging.

File: extensions/shared/register-jetpack-block.js [hunk 3]
```diff
@@ -48,6 +50,14 @@ export default function registerJetpackBlock( name, settings, childBlocks = [] )
 		edit: requiredPlan ? wrapPaidBlock( { requiredPlan } )( settings.edit ) : settings.edit,
 	} );
 
+	if ( requiredPlan ) {
+		addFilter(
+			'editor.BlockListBlock',
+			`jetpack/${ name }-with-has-warning-is-interactive-class-names`,
```
Comment (simison) [inline extensions/shared/register-jetpack-block.js]: Haha that's a mouthful!

File: extensions/shared/with-has-warning-is-interactive-class-names/style.scss (hunk 2 above)
Comment (simison) [reply extensions/shared/with-has-warning-is-interactive-class-names/style.scss]: I see it's explained in https://github.com/Automattic/jetpack/pull/13019/files#diff-37c66df99b9d7cba3dc6798b7e53d938R11 but might be worth reiterating nevertheless.

**Review by simison — DISMISSED** [review]
Looking good and works fabulously well!

File: extensions/shared/with-has-warning-is-interactive-class-names/index.jsx [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+// Injecting the `has-warning` class into the block wrapper component gives us
+// the right kind of borders around the block, both visually and conceptually.
+// However, it also adds styling to prevent user interaction with that block.
+// We thus add a new `is-interactive` class to be able to override that behavior.
+export default name =>
+	createHigherOrderComponent(
+		BlockListBlock => props => (
+			<BlockListBlock
+				{ ...props }
+				className={ props.name === name ? 'has-warning is-interactive' : '' }
```
Comment (simison) [inline extensions/shared/with-has-warning-is-interactive-class-names/index.jsx]: `is-interactive` isn't a core class, right? Therefore I'd jetpack-prefix it. Not a blocker tho.

File: extensions/shared/with-has-warning-is-interactive-class-names/style.scss (hunk 2 above)
Comment (simison) [reply extensions/shared/with-has-warning-is-interactive-class-names/style.scss]: Added a note in https://github.com/Automattic/jetpack/pull/13019/commits/5f9b77aea9b8bb49ef98101338d566a1b2f1c760

File: extensions/shared/register-jetpack-block.js (hunk 3 above)
Comment (ockham) [reply extensions/shared/register-jetpack-block.js]: Yeah, same for the component name :grimacing: Happy to change it something more palatable!

**Review by zinigor — APPROVED** [review]
Works well for me, the border appears as it should.

**Pull request merged**